### ISSUE TITLE
Add fractional flow constraints in allocation

### ISSUE THREAD

**Issue by SouthEndMusic** [Deltares/Ribasim#727]
**What**
The optimizations for allocation flows must take `FractionalFlow` nodes into account as a constraint.

**Why**
The constrains posed within the allocation optimization problems make the optimizer reasonably aware of how water can flow from sources to users. The more accurate this is, the more accurately water can be allocated, so that the flow rate allocated to a user is close to the users' actual abstraction rate.

**How**
The work to be done can be divided into several parts:
- [x] Constraints must be added to the problem within the `JuMP` problem constructor function `allocation_problem`. this can be done analogous to the existing `add_constraints_*` functions in `allocation.jl`.
- [x] This could maybe be done in a follow-up PR: currently there are 'conservation' constraints: for each node the summed outflow is bounded from above by the summed inflow, so that the optimizer cannot 'make up' water. The fractional flow constraint will make this redundant for nodes that have fractional flow outflows so these constraints can be removed. _Edit: after reconsideration this turns out not to be necessary, since these conservation constraints are on basins, and the fractional flow constraints are on flow prescribing nodes._
- [x] Since https://github.com/Deltares/Ribasim/pull/716 `FractionalFlow` nodes are also controllable, meaning that their fractions can change over simulation time. Allocation has to be made aware of these changes. I think the best place to do this is in the same place where the parameters of the main simulations are updated; in `set_control_params!` in `bmi.jl`. 

**Additional context**
- I've noticed that `FractionalFlow` nodes aren't documented that well, there's only this: https://deltares.github.io/Ribasim/core/usage.html#fractionalflow. No entry on the equations page.
- For a dive into the theory of allocation see https://deltares.github.io/Ribasim/core/allocation.html.

**Comment by SouthEndMusic** [comment]
@deltamarnix after refining the issue it looks more like a deep dive than a beginners issue for allocation 😋 Things are quite interrelated

**Comment by SouthEndMusic** [comment]
Also, consider this issue blocked by https://github.com/Deltares/Ribasim/pull/807, that refactor will make allocation development much easier.

**Comment by SouthEndMusic** [comment]
@visr for water conservation there currently aren't equality constraints; the total outflow of a node is bounded from above by the total inflow of the node. This means that water can be 'lost' at a node, but not gained, which I think is fine for the purposes of allocation. Also, equality constraints are probably harder for the solver to satisfy.

This brings me to the constraints that must be introduced for fractional flow. Is it okay to do those in a similar way, i.e. the flow after the fractional flow node is bounded from above by the inflow times the fraction? If it is done in that way, the flow after the fractional flow nodes does generally not strictly adhere to the fractions anymore. As I see it, it depends on the physical interpretation of the fractional flow node whether this is okay or not.

**Comment by SouthEndMusic** [comment]
The comment above was discussed, and the proposed constraints are fine until proven later that the deviation between allocation flow and physical flow is too large.

**Comment by SouthEndMusic** [comment]
@visr I can imagine that there are situations where the above does not go well if there is flow over the fractional flow nodes for multiple priorities. The problem here is that the fractional flow constraint should hold for the total flow over these nodes, not the flow per priority as we do with the goal-programming approach for allocation. We'll see how this turns out in practice.
